launcher/agent/agent.go

```diff
@@ -115,25 +154,71 @@ func (a *agent) Attest(ctx context.Context, opts AttestAgentOpts) ([]byte, error
 		return nil, fmt.Errorf("failed to get principal tokens: %w", err)
 	}
 
-	var buf bytes.Buffer
-	if err := a.cosCel.EncodeCEL(&buf); err != nil {
-		return nil, err
-	}
-
-	attestation, err := a.attest(challenge.Nonce, buf.Bytes())
-	if err != nil {
-		return nil, fmt.Errorf("failed to attest: %v", err)
-	}
-
 	req := verifier.VerifyAttestationRequest{
 		Challenge:      challenge,
 		GcpCredentials: principalTokens,
-		Attestation:    attestation,
+		Attestation:    nil,
 		TokenOptions: verifier.TokenOptions{
 			CustomAudience: opts.Aud,
 			CustomNonce:    opts.Nonces,
 			TokenType:      opts.TokenType,
 		},
+		TDCCELAttestation: nil,
+	}
+
+	attResult, err := a.ea.Attest(challenge.Nonce)
+	if err != nil {
+		return nil, fmt.Errorf("failed to attest: %v", err)
+	}
+
+	switch v := attResult.(type) {
+	case *pb.Attestation:
+		a.logger.Println("attestation through TPM quote")
+
+		var cosCelPcr bytes.Buffer
+		if err := a.cosCel.EncodeCEL(&cosCelPcr); err != nil {
+			return nil, err
+		}
+
+		v.CanonicalEventLog = cosCelPcr.Bytes()
+		req.Attestation = v
+	case *tdxpb.QuoteV4:
+		a.logger.Println("attestation through TDX quote")
+
+		ccelData, err := os.ReadFile("/sys/firmware/acpi/tables/data/CCEL")
+		if err != nil {
+			return nil, err
+		}
+		ccelTable, err := os.ReadFile("/sys/firmware/acpi/tables/CCEL")
+		if err != nil {
+			return nil, err
+		}
+
+		var cosCelRtmr bytes.Buffer
+		if err := a.cosCel.EncodeCEL(&cosCelRtmr); err != nil {
+			return nil, err
+		}
+
+		certChain, err := a.fetchedAK.GetCertificateChain(http.DefaultClient)
+		if err != nil {
+			return nil, fmt.Errorf("failed when fetching certificate chain: %w", err)
+		}
+
+		rawQuote, err := tabi.QuoteToAbiBytes(v)
+		if err != nil {
+			return nil, err
+		}
+		req.TDCCELAttestation = &verifier.TDCCELAttestation{
+			CcelAcpiTable:     ccelTable,
+			CcelData:          ccelData,
+			CanonicalEventLog: cosCelRtmr.Bytes(),
+			TdQuote:           rawQuote,
+			// remove following two fields once found other way to get GCE info
+			AkCert:            a.fetchedAK.CertDERBytes(),
+			IntermediateCerts: certChain,
+		}
+	default:
+		return nil, fmt.Errorf("received an unsupported attestation type! %v", v)
```

Comment:
Could you extract this entire logic to a new function something like `determineReqAttestation`?

Reply:
it's moved to `func (t *tdxAttestRoot) Attest(nonce []byte)`